implementations/rust/ockam/ockam_api/src/http_auth/mod.rs

```diff
@@ -0,0 +1,366 @@
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use ockam::errcode::{Kind, Origin};
+
+use tracing::{debug, error};
+
+use crate::TokenLeaseRefresher;
+
+#[derive(Debug, Clone, PartialEq)]
+enum RequestState {
+    ParsingHeader(Option<Vec<u8>>),
+    ParsingChunkedHeader(Option<Vec<u8>>),
+    RemainingInChunk(usize),
+    RemainingBody(usize),
+}
+
+struct HttpAuthInterceptorState {
+    state: RequestState,
+}
+
+struct HttpAuthInterceptor {
+    state: Arc<Mutex<HttpAuthInterceptorState>>,
+    token_refresher: TokenLeaseRefresher,
+}
+
+impl HttpAuthInterceptor {
+    fn new(token_refresher: TokenLeaseRefresher) -> Self {
+        let state = HttpAuthInterceptorState {
+            state: RequestState::ParsingHeader(None),
+        };
+        Self {
+            state: Arc::new(Mutex::new(state)),
+            token_refresher,
+        }
+    }
+}
+
+pub struct HttpAuthInterceptorFactory {
+    token_refresher: TokenLeaseRefresher,
+}
+
+impl HttpAuthInterceptorFactory {
+    pub fn new(token_refresher: TokenLeaseRefresher) -> Self {
+        Self { token_refresher }
+    }
+}
+
+impl PortalInterceptorFactory for HttpAuthInterceptorFactory {
+    fn create(&self) -> Arc<dyn PortalInterceptor> {
+        Arc::new(HttpAuthInterceptor::new(self.token_refresher.clone()))
+    }
+}
+
+fn attach_auth_token_and_serialize_into(
+    req: &httparse::Request,
+    token: &str,
+    buffer: &mut Vec<u8>,
+) {
+    debug!("Serializing http req header");
+    write!(
+        buffer,
+        "{} {} HTTP/1.{}\r\n",
+        req.method.unwrap(),
+        req.path.unwrap(),
+        req.version.unwrap()
+    )
+    .unwrap();
+
+    write!(buffer, "Authorization: Token {}\r\n", token).unwrap();
+    for h in &*req.headers {
+        if !h.name.eq_ignore_ascii_case("Authorization") {
+            write!(buffer, "{}: ", h.name).unwrap();
+            buffer.extend_from_slice(h.value);
+            buffer.extend_from_slice(b"\r\n");
+        }
+    }
+    buffer.extend_from_slice(b"\r\n");
+}
+
+fn body_state(method: &str, headers: &[Header]) -> ockam_core::Result<RequestState> {
+    if ["POST".to_string(), "PUT".to_string()].contains(&method.to_uppercase()) {
+        for h in headers {
+            if h.name.eq_ignore_ascii_case("Content-Length") {
+                if let Ok(str) = std::str::from_utf8(h.value) {
+                    return str
+                        .parse()
+                        .map(RequestState::RemainingBody)
+                        .map_err(|e| ockam_core::Error::new(Origin::Transport, Kind::Invalid, e));
+                }
+            } else if h.name.eq_ignore_ascii_case("Transfer-Encoding")
+                && String::from_utf8(h.value.to_vec()).is_ok_and(|s| s.contains("chunked"))
+            {
+                return Ok(RequestState::ParsingChunkedHeader(None));
+            }
+        }
+        // Not content-length, no chunked encoding, fail.
+        Err(ockam_core::Error::new(
```

Comment:
ha, 👍  will incorportate it anyway.  I didn't encounter it when testing with influx, but who knows

Reply:
hum,  but these are _requests_ only, and `Connection: close` seems a response-only thing

>  5.By the server closing the connection. (Closing the connection
     cannot be used to indicate the end of a request body, since that
     would leave no possibility for the server to send back a response.)

will let it for latter,  _maybe_ at some point we need to parse responses too (at that point need to adapt other parts of the code as well)